src/clr-angular/data/datagrid/datagrid-pagination.ts

```diff
@@ -3,58 +3,45 @@
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
-import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
+import { Component, ContentChild, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
 import { Subscription } from 'rxjs';
-
 import { Page } from './providers/page';
+import { ClrDatagridPageSize } from './datagrid-page-size';
 
 @Component({
   selector: 'clr-dg-pagination',
   template: `
-        <div class="pagination-description">
-            <ng-content></ng-content>
-        </div>
-        <ul class="pagination-list" *ngIf="page.last > 1">
-            <li *ngIf="page.current > 1">
-                <button
-                        class="pagination-previous"
-                        (click)="page.previous()"
-                        type="button"></button>
-            </li>
-            <li *ngIf="page.current > 2">
-                <button (click)="page.current = 1" type="button">1</button>
-            </li>
-            <li *ngIf="page.current > 3">...</li>
-            <li *ngFor="let pageNum of middlePages" [class.pagination-current]="pageNum === page.current">
-                <button
-                        *ngIf="pageNum !== page.current; else noButton"
-                        (click)="page.current = pageNum"
-                        type="button">{{pageNum}}
-                </button>
-                <ng-template #noButton>{{pageNum}}</ng-template>
-            </li>
-            <li *ngIf="page.current < page.last - 2">...</li>
-            <li *ngIf="page.current < page.last - 1">
-                <button
-                        (click)="page.current = page.last"
-                        type="button">{{page.last}}
-                </button>
-            </li>
-            <li *ngIf="page.current < page.last">
-                <button
-                        class="pagination-next"
-                        (click)="page.next()"
-                        type="button"></button>
-            </li>
-        </ul>
+    <div class="pagination-size" *ngIf="pageSizeComponent">
+      <ng-content select="clr-dg-page-size"></ng-content>
+    </div>
+    <div class="pagination-description">
+      <ng-content></ng-content>
+    </div>
+    <div class="pagination-list" *ngIf="page.last > 1">
+      <button class="pagination-first" [disabled]="page.current === 1" (click)="page.current = 1">
+        <clr-icon shape="angle-double left"></clr-icon>
+      </button>
+      <button class="pagination-previous" [disabled]="page.current === 1" (click)="page.current = page.current - 1">
+        <clr-icon shape="angle left"></clr-icon>
+      </button>
+      <input type="text" class="pagination-current" [size]="page.last.toString().length" [(ngModel)]="page.current" />&nbsp;/&nbsp;<span>{{page.last}}</span>
```

Comment:
And... 🌮. I could see this comment requiring large changes. 😢 

You're using `[(ngModel)]` on a text input. 
1. What happens if the user types "hello"?
2. What happens if the user types a negative number? A page that's after the last page? I don't believe the pagination service puts guards against this.
3. This is the scary one: If I want to go to page 123, `[(ngModel)]` fires **as I type**. So it will move to page 1, then 12, then finally 123. If each of these triggers an expensive HTTP call to the server, it's bad. Not to mention the fact that if people didn't implement their fetch with proper cancellation, you'll have race conditions and might end up on page 123 with the items of page 12. Admittedly that's the app's fault, but I'd rather avoid having to document how to properly fire HTTP requests to the server and cancel them when they are in quick succession.

Reply:
1 and 2: It doesn't blow up, but the Page service will set it to whatever it's passed. We can use the input `type=number` and play around with css for calculating proper width (although I think Edge didn't support `type=number` perfectly?). So maybe a better approach is to enforce validation at the component-level (e.g. keyup event?) and also add a safeguard to the Page's set current method. Thoughts?

3: Instead of using ngModel could we use a custom input/output here, and we use debounce to ignore rapid succession of changes in input value when we emit the output?

Reply:
For 3, I would ignore everything until the user presses Enter or blurs the input.

This means that for 1 and 2, I would probably let the user type stupid things, and fallback on page 1 if they enter an invalid value (or on the last page if they enter a value greater than the number of pages).